From e8a99cbfd755b890dece6a4412487f9621c7d59c Mon Sep 17 00:00:00 2001
From: Ky-Anh Huynh <kyanh@theslinux.org>
Date: Tue, 30 Apr 2013 15:07:18 +0700
Subject: [PATCH] + config. to set default mode for generic IM

New configuration option allows user to select default mode
for generic IM. This is useful but it may be buggy if some IM
has its own configuration to do the samething.

FIXME: Add test cases
---
 uim-vi/src/uim-1.8.5/scm/generic-custom.scm |  6 ++++++
 uim-vi/src/uim-1.8.5/scm/generic.scm        | 10 ++++++++--
 uim-vi/src/uim-1.8.5/scm/im-custom.scm      |  5 +++++
 3 files changed, 19 insertions(+), 2 deletions(-)

diff --git a/uim-vi/src/uim-1.8.5/scm/generic-custom.scm b/uim-vi/src/uim-1.8.5/scm/generic-custom.scm
index 0b6811d..4505076 100644
--- a/uim-vi/src/uim-1.8.5/scm/generic-custom.scm
+++ b/uim-vi/src/uim-1.8.5/scm/generic-custom.scm
@@ -36,6 +36,12 @@
 		     (N_ "Other input methods")
 		     (N_ "long description will be here."))
 
+(define-custom 'generic-im-default-state-on? #f
+  '(other-ims miscellaneous)
+  '(boolean)
+  (N_ "Default input mode is on")
+  (N_ "long description will be here."))
+
 (define-custom 'generic-use-candidate-window? #t
   '(other-ims candwin)
   '(boolean)
diff --git a/uim-vi/src/uim-1.8.5/scm/generic.scm b/uim-vi/src/uim-1.8.5/scm/generic.scm
index d632c10..e287756 100644
--- a/uim-vi/src/uim-1.8.5/scm/generic.scm
+++ b/uim-vi/src/uim-1.8.5/scm/generic.scm
@@ -40,14 +40,17 @@
 (define generic-widgets '(widget_generic_input_mode))
 
 ;; default activity for each widgets
-(define default-widget_generic_input_mode 'action_generic_off)
+(define default-widget_generic_input_mode
+  (cond
+    (generic-im-default-state-on? 'action_generic_on)
+    (else 'action_generic_off)
+  ))
 
 ;; actions of widget_generic_input_mode
 (define generic-input-mode-actions
   '(action_generic_off
     action_generic_on))
 
-
 ;;; implementations
 
 (define ascii-rule
@@ -61,6 +64,9 @@
   (lambda (gc)
     (let ((rkc (generic-context-rk-context gc)))
       (rk-flush rkc)
+      (cond
+        (generic-im-default-state-on? 'action_generic_on)
+        (else 'action_generic_off))
       (generic-update-preedit gc))))
 
 (register-action 'action_generic_off
diff --git a/uim-vi/src/uim-1.8.5/scm/im-custom.scm b/uim-vi/src/uim-1.8.5/scm/im-custom.scm
index fed60d4..3b161a6 100644
--- a/uim-vi/src/uim-1.8.5/scm/im-custom.scm
+++ b/uim-vi/src/uim-1.8.5/scm/im-custom.scm
@@ -86,6 +86,11 @@
 		     (N_ "long description will be here."))
 
 ;; subgroup
+(define-custom-group 'miscellaneous
+		     (N_ "Miscellaneous settings")
+		     (N_ "long description will be here."))
+
+;; subgroup
 (define-custom-group 'dictionary
                      (N_ "Dictionary")
                      (N_ "long description will be here."))
-- 
1.8.2.1

